### PR TITLE
Create package async-registry

### PR DESCRIPTION
**Pull request by happykmm**
Create package async-registry

### CODE REVIEW

File: packages/async-registry/src/index.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+class AsyncStrategy extends Strategy {
+  get = name => (this.registry.has(name) ?
+    Promise.resolve(this.registry.get(name)()) :
+    Promise.resolve(null))
+}
+
+
+export default class AsyncRegistry {
+  strategies = List([
```
Comment (wewei) [inline packages/async-registry/src/index.js]:
```js
defaultStrategy = new Strategy()
defaultAsyncStrategy = new AsyncStrategy()

strategies = List([
  defaultStrategy,
  defaultAsyncStrategy,
]);
```

File: packages/async-registry/src/index.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import _ from 'lodash';
+import { List, Map } from 'immutable';
+import Promise from 'bluebird';
+
+Promise.config({
+  cancellation: true,
+});
```
Comment (wewei) [inline packages/async-registry/src/index.js]: Shouldn't be here

File: projects/atlas/new-src/ui/components/cell.js [hunk 3]
```diff
@@ -2,24 +2,51 @@ import React, { PureComponent } from 'react';
...
+    }
+  }
+
+  updateControl(type, isCtor = false) {
+    const controlPromise = componentRegistry.get(type);
+    if (controlPromise.isFulfilled()) {
+      // No need to wait for next tick
+      const Control = controlPromise.value();
+      if (isCtor) {
+        this.state = _.defaults({
```
Comment (wewei) [inline projects/atlas/new-src/ui/components/cell.js]: Move this to ctor

File: projects/atlas/new-src/ui/components/cell.js [hunk 4]
```diff
@@ -2,24 +2,51 @@ import React, { PureComponent } from 'react';
...
+
   state = {
     Control: _.constant(null),
   }
 
   componentWillReceiveProps(nextProps) {
-    const Control = ComponentRegistry.get(nextProps.type);
-    if (Promise.resolve(Control) === Control) { // Get an async compoennt
-      Control.then((AsyncControl) => {
+    if (nextProps.type !== this.props.type) {
```
Comment (wewei) [inline projects/atlas/new-src/ui/components/cell.js]: Use `getDerivedStateFromProps`

**Review by wewei — CHANGES_REQUESTED** [review]

File: packages/async-registry/src/index.js (hunk 1 above)
Comment (happykmm) [reply packages/async-registry/src/index.js]: Resolve

File: packages/async-registry/src/index.js (hunk 2 above)
Comment (happykmm) [reply packages/async-registry/src/index.js]: Resolve

**Review by wewei — APPROVED** [review]

**Pull request merged**